// @psychedelic/plug-inpage-provider/dist/src/utils/sign.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArgTypes = exports.signFactory = exports.decodeArgs = exports.getSignInfoFromTransaction = exports.canDecodeArgs = void 0;
const candid_1 = require("@dfinity/candid");
const buffer_1 = require("buffer/");
const bigint_1 = require("./bigint");
const domain_metadata_1 = __importDefault(require("./domain-metadata"));
const canDecodeArgs = (signInfo, argsTypes) => {
    var _a;
    return !!((signInfo === null || signInfo === void 0 ? void 0 : signInfo.canisterId) &&
        (signInfo === null || signInfo === void 0 ? void 0 : signInfo.methodName) &&
        (signInfo === null || signInfo === void 0 ? void 0 : signInfo.arguments) &&
        ((_a = argsTypes[signInfo.canisterId]) === null || _a === void 0 ? void 0 : _a[signInfo.methodName]));
};
exports.canDecodeArgs = canDecodeArgs;
const getSignInfoFromTransaction = (transaction, sender) => {
    var _a;
    const interfaceFactory = transaction.idl({ IDL: candid_1.IDL });
    const [methodName, func] = ((_a = interfaceFactory === null || interfaceFactory === void 0 ? void 0 : interfaceFactory._fields) === null || _a === void 0 ? void 0 : _a.find(([methodName, _func]) => methodName === transaction.methodName)) || [undefined, undefined];
    const decodedArguments = Array.isArray(transaction.args)
        ? transaction.args
        : undefined;
    return {
        methodName: methodName || transaction.methodName,
        canisterId: transaction.canisterId,
        sender,
        arguments: decodedArguments && Array.isArray(transaction.args) && func
            ? candid_1.IDL.encode(func.argTypes, transaction.args)
            : buffer_1.Buffer.from([]),
        decodedArguments,
        preApprove: false,
        requestType: "unknown",
    };
};
exports.getSignInfoFromTransaction = getSignInfoFromTransaction;
const decodeArgs = (signInfo, argsTypes) => {
    if ((0, exports.canDecodeArgs)(signInfo, argsTypes)) {
        const assuredSignInfo = signInfo;
        const funArgumentsTypes = argsTypes[assuredSignInfo.canisterId][assuredSignInfo.methodName];
        return candid_1.IDL.decode(funArgumentsTypes, assuredSignInfo.arguments);
    }
};
exports.decodeArgs = decodeArgs;
const signFactory = (clientRPC, argsTypes, preApprove = false) => (payload, signInfo) => __awaiter(void 0, void 0, void 0, function* () {
    const metadata = (0, domain_metadata_1.default)();
    const payloadArr = new Uint8Array(payload);
    if (signInfo)
        signInfo.decodedArguments = signInfo.arguments
            ? (0, bigint_1.recursiveParseBigint)((0, exports.decodeArgs)(signInfo, argsTypes))
            : [];
    const res = yield clientRPC.call({
        handler: "requestSign",
        args: [payloadArr, metadata, Object.assign(Object.assign({}, signInfo), { preApprove })],
    });
    return new Uint8Array(Object.values(res));
});
exports.signFactory = signFactory;
const getArgTypes = (interfaceFactory) => {
    const service = interfaceFactory({ IDL: candid_1.IDL });
    const methodArgType = {};
    service._fields.forEach(([methodName, fun]) => (methodArgType[methodName] = fun.argTypes));
    return methodArgType;
};
exports.getArgTypes = getArgTypes;
